### ISSUE TITLE
Discontinuities in saved data with AXI DMA in cyclic Scatter/Gather Mode

### ISSUE THREAD

**Issue by rjmigliori** [OpenDGPS/zynq-axi-dma-sg#4]
HI OpenDGPS, 
I'm trying to save AXI4 stream data to memory using the AXI DMA engine on a Red Pitaya, which has I believe a Zynq 7010 SoC. 

I've had some luck developing a project to do this based on your code, but there are strange and inexplicable discontinuities in the data saved to memory by the DMA engine, as in this [figure](https://drive.google.com/open?id=1q1RRTDRAIU58SDWv9QjSpiGQ4syPf4Ad). 

In order to discern the nature of these discontinuities, I built a vivado program which does the following 

- Writes the output of a binary counter to the AXI4 stream. 
- Passes AXI4 Stream to a FIFO 
- FIFO writes the AXI4 stream to the DMA engine 

Here is the corresponding [block diagram](https://drive.google.com/open?id=1jbZGkbUyB1ND5xIsKhcfGVzBJcenAdZx), and [address editor](https://drive.google.com/open?id=1ICKQwYlSNOf8GqgnQCYxgxV56WQ5O3Ex). 

Based on your code, I wrote a c program [HP0_DMA_SG.c](https://drive.google.com/open?id=1M7dJsvqVDjOKc0DlZlPJUuWrZz_ykne8) that does the following

* Define Variables and functions 
* Create Memory Maps 
    * axi_dma 
    * mm2s and s2mm descriptor chain 
    * source memory (mm2s) 
    * destination memory (s2mm) 
* Clear descriptor chains, destination memory 
* reset DMA engine 
* Build descriptor chain 
    * First Descriptor points to second descriptor, transfers TRANSFER_BYTES bytes to destination address, has complete frame flag set
    * Second descriptor points to third descriptor, transfers TRANSFER_BYTES bytes to destination address + TRANSFER_BYTES, has complete frame flag set. 
    * Third descriptor points to fourth descriptor, transfers TRANSFER_BYTES bytes to destination address+2*TRANSFER_BYTES, has complete frame flag set 
    * Fourth descriptor points to first descriptor, transfers TRANSFER_BYTES bytes to destination address +3*TRANSFER_BYTES, has complete frame flag set 
* Initiate a cyclic S2MM transfer by setting the S2MM tail descriptor accordingly
* Print out a few statuses of the s2mm descriptors 
* Halt and reset the DMA 
* Pass data written to destination address by the descriptor chain through a created socket to a python client. 

The python client determines where (i.e. at what index) the discontinuities are in the received data and the computes the spacing between them. **The distance between discontinuities appears to be a function of, and sometimes equivalent to, the number of bytes transferred by each descriptor** 

[Here](https://drive.google.com/open?id=1qRFG6alBEFEQKp73_RyPhhh0W4H6DwJq) is a plot of the received data, with discontinuities indicated as red dots/vertical lines, in a test where each descriptor was to transfer 1024 bytes (256 integers). Also notice in this figure where the signal drops to zero - I believe this is because the fourth descriptor is for some reason not writing to memory. 

In the provided figure above, the red dots indicate the positions of discontinuities. Here are their positions and the spacing between them: 
```
Discontinuity index, Discontinuity spacing 
249
250
505, 505-249=256
506
761, 761-505=256
762
767, 767-761=6
768
```
Notice that some of the discontinuities are 256 integers, or 1024 bytes, apart. 

Additionally, I believe that the stream (or something) is not being flushed properly when I reset the DMA engine. I.e., I think the BD chain is retaining data after, even after I halt and reset it. I'll illustrate this with an example

- Suppose I run a test where each descriptor is to transfer 1024 bytes. This results in a figure such as [this](https://drive.google.com/open?id=1l92o7PC0M33WNPM05PRybhGUb2I-PlpR)
- If I follow that immediately with a test where each descriptor is to transfer 8192 bytes, the result is a figure such as [this](https://drive.google.com/open?id=1Bn3cs7fVz1jBRi3A4m77JN3e-T8cbcRO). Notice how the first (I'm guessing exactly) 1024 samples have discontinuities spaces as though each descriptor was transferring 1024 bytes, then the discontinuities spread out after that point. 

I would really appreciate some help debugging this discontinuity issue. I understand that this is a lot of information to process. Please let me know if anything I've said deems further clarification or explanation, I'm happy to answer any questions you have. 

Thanks!

**Comment by abuzarshah963** [comment]
Hi rjmigliori,
I am also working on same thing. I have achieved continuous samples from PL to PS (MM2S and S2MM) and successfully transferred over ethernet. 
You have to make an application with "LWIP echo server" and add the attached three files. Then using putty (baud rate 115200, COM3) you will get the data by uncommenting line **"//xil_printf("%d,",*impval);"** in **echo.c**
Else you can detect live data on matlab or labview by using TCP protocol with board IP "192.168.1.10", subnet mask "255.255.255.0" and GW "192.168.1.1"

Note: "I have sent sine wave of 5MHz from zedboard over ethernet server"

![image](https://user-images.githubusercontent.com/49023168/65502907-ab20b500-de78-11e9-827b-ad7abbc5cb09.png)

[code_Eth.zip](https://github.com/OpenDGPS/zynq-axi-dma-sg/files/3646588/code_Eth.zip)

**Comment by abuzarshah963** [comment]
"LWIP echo server" in SDK (Vivado 2018.3 preferably).

**Comment by huzaifasajids** [comment]
Hi Abuzar, I'm working on exactly the same project as yours. Would love to get some help. If you can, please contact me through email(mentioned on my profile). Thanks.

**Comment by abuzarshah963** [comment]
> Hi Abuzar, I'm working on exactly the same project as yours. Would love to get some help. If you can, please contact me through email(mentioned on my profile). Thanks.

I have attached zip file with my comment. Refer to my reply.

**Comment by huzaifasajids** [comment]
yeah, I saw that. That's why I am contacting you now. Would have been excellent if you added the vivado design / simulink design to the zip as well. I'm trying to run your code but am probably making some mistake in the design.

**Comment by rjmigliori** [comment]
@abuzarshah963 I'll have to take a look at this at some point. Also the vivado design would certainly be useful to see as well. Thanks for replying!

**Comment by abuzarshah963** [comment]
> yeah, I saw that. That's why I am contacting you now. Would have been excellent if you added the vivado design / simulink design to the zip as well. I'm trying to run your code but am probably making some mistake in the design.

Block diagram is the same as attached with hyperlink by rjmigliori.

**Comment by abuzarshah963** [comment]
> > yeah, I saw that. That's why I am contacting you now. Would have been excellent if you added the vivado design / simulink design to the zip as well. I'm trying to run your code but am probably making some mistake in the design.
> 
> Block diagram is the same as attached with hyperlink by rjmigliori.

If still facing any problem, contact me via email. I'll share the design details with you.
